### PR TITLE
[FEATURE] Store Custom Fields and Remove from PluginStore #169494035

### PR DESCRIPTION
**Pull request by dannyb**
# Overall

Stores the `partreon_id`, `has_nsfw_campaign` in the `user.custom_fields` and removes the data from the `PluginStore` after the user has been created.

## Finishes

* [#169494035](https://www.pivotaltracker.com/story/show/169494035)

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -31,23 +31,58 @@ def self.default_image_url
...
+        user.custom_fields["has_nsfw_campaign"] = has_nsfw_campaign
+        user.save_custom_fields
+
+        # Remove the temp data from the plugin store
+        Patreon.remove(plugin_store_key)
+      else
+        # User has not been created, but they are authorized.
+        # save data in plugin store so we can use in during
+        # This is called during authorization and on the
+        # :user_created callback.
```
Comment (elskwid) [inline plugin.rb]: This comment seems like it either isn't complete or may be two thoughts. 

> so we can use in during This is called 

just reads weird

File: plugin.rb [hunk 2]
```diff
@@ -31,23 +31,58 @@ def self.default_image_url
...
     end
+
+    def self.store
+      @store ||= PluginStore.new(PLUGIN_NAME)
+    end
+
+    def self.add_remove_nsfw_group(user, has_nsfw_campaign)
+      return unless user && Patreon.nsfw_group
+
+      has_nsfw_campaign ? Patreon.nsfw_group.add(user) : Patreon.nsfw_group.remove(user)
```
Comment (elskwid) [inline plugin.rb]: Is adding a user to a group idempotent?

File: plugin.rb [hunk 3]
```diff
@@ -120,17 +155,16 @@ def raw_info
   DiscourseEvent.on(:user_created) do |user|
     if SiteSetting.patreon_creator_enabled
       begin
-        nsfw_group = Group.find_by_name(SiteSetting.patreon_creator_nsfw_group)
-
-        user_info = Patreon.get("user_info") || {}
-        user_record = user_info[user.email] || {}
-        has_nsfw_campaign = user_record[:has_nsfw_campaign]
+        user_info_record = Patreon.get("user_info.#{user.email}") || {}
```
Comment (elskwid) [inline plugin.rb]: Non-blocking: I could see `Patreon.get("user_info.#{user.email}")` moving to a helper like 

```ruby
def get_user_info_record(email)
    key = "user_info.#{email}"
    Patreon.get(key) || {}
end
```

File: plugin.rb [hunk 4]
```diff
@@ -172,23 +206,20 @@ def after_authenticate(auth_token)
...
-      user_info[result.email] = user_record
-      Patreon.set("user_info", user_info)
-
-      if nsfw_group && user
-        has_nsfw_campaign ? nsfw_group.add(user) : nsfw_group.remove(user)
-      end
+      Patreon.add_remove_nsfw_group(user, has_nsfw_campaign)
+
+      # Update the user or store the data in the
+      # Plugin store.  This also used in the
```
Comment (elskwid) [inline plugin.rb]: This comment looks incomplete.

**Review by elskwid — CHANGES_REQUESTED** [review]
The comments seem like there might be a little chopped up. Clean those up and this is goog to go.

I like the refactoring.

File: plugin.rb (hunk 2 above)
Comment (dannyb) [reply plugin.rb]: @elskwid yes, `Group.add` and `Group.remove` are idempotent.

https://github.com/discourse/discourse/blob/master/app/models/group.rb#L614
https://github.com/discourse/discourse/blob/master/app/models/group.rb#L641

File: plugin.rb (hunk 3 above)
Comment (dannyb) [reply plugin.rb]: yeah I thought about it, I will do it too

File: plugin.rb (hunk 2 above)
Comment (elskwid) [reply plugin.rb]: Great, thanks.

**Review by elskwid — APPROVED** [review]

**Comment by elskwid** [comment]
@dannyb I've marked this as approved so after the comments are updated you can merge away.

**Pull request merged**